File: job-server-extras/src/main/java/spark/jobserver/examples/StreamingTestJob.java
package spark.jobserver.examples;

import com.typesafe.config.Config;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.streaming.StreamingContext;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaPairDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import spark.jobserver.api.JobEnvironment;
import spark.jobserver.api.JobValidation;
import spark.jobserver.context.JSparkStreamingJob;

import java.util.Arrays;
import java.util.LinkedList;

public class StreamingTestJob extends JSparkStreamingJob<Void> {
    @Override
    public Void runJob(StreamingContext context, JobEnvironment cfg, Config data) {
        final JavaStreamingContext jsc = new JavaStreamingContext(context);
        final JavaRDD<Integer> d = jsc.sparkContext().parallelize(Arrays.asList(1, 2, 3, 4, 5));

        final LinkedList<JavaRDD<Integer>> q = new LinkedList<>();
        q.add(d);

        final JavaDStream<Integer> dStream = jsc.queueStream(q);
        final JavaPairDStream<Integer, Long> counts = dStream.countByValue();

        counts.print(5);
        jsc.start();
        jsc.awaitTermination();
        return null;
    }

    @Override
    public JobValidation validate(StreamingContext context, JobEnvironment jEnv, Config cfg) {
        return new JobValidation.JOB_VALID();
    }
}

File: job-server-extras/src/main/scala/spark/jobserver/StreamingTestJob.scala
package spark.jobserver

import com.google.common.annotations.VisibleForTesting
import com.typesafe.config.Config
import org.apache.spark.rdd.RDD
import org.apache.spark.streaming.StreamingContext

import scala.collection.mutable
import scala.util.Try

@VisibleForTesting
object StreamingTestJob extends SparkStreamingJob {
  def validate(ssc: StreamingContext, config: Config): SparkJobValidation = SparkJobValid

  def runJob(ssc: StreamingContext, config: Config): Any = {
    val maxIterations = Try(config.getInt("streaming.test.job.maxIterations")).getOrElse(1)
    val totalDelaySeconds = Try(config.getInt("streaming.test.job.totalDelaySeconds")).getOrElse(0)
    val shouldPrintCount = Try(config.getBoolean("streaming.test.job.printCount")).getOrElse(true)

    val queue = mutable.Queue[RDD[String]]()
    val lines = ssc.queueStream(queue)
    val words = lines.flatMap{ l =>
      Thread.sleep(totalDelaySeconds * 1000)
      l.split(" ")
    }

    val wordCounts = words.countByValue()
    //do something
    wordCounts.foreachRDD(rdd =>
    try {
        shouldPrintCount match {
          case true => println(rdd.count())
          case false => rdd.count()
        }
      } catch {
        case _: InterruptedException => {}
      })
    ssc.start()
    var totalIterations = 0
    while(totalIterations < maxIterations) {
      queue.synchronized {
        queue += ssc.sparkContext.makeRDD(Seq("123", "test", "test2"))
      }
      totalIterations += 1
      Thread.sleep(100)
    }
    ssc.awaitTermination()
  }
}
